src/zope/server/trigger.py

```diff
@@ -123,106 +120,126 @@ def handle_read(self):
             self.recv(8192)
         except socket.error:
             return
-        self.lock.acquire()
-        try:
+        with self.lock:
             for thunk in self.thunks:
                 try:
                     thunk()
                 except:
-                    nil, t, v, tbinfo = asyncore.compact_traceback()
+                    _nil, t, v, tbinfo = asyncore.compact_traceback()
                     print('exception in trigger thunk:'
                           ' (%s:%s %s)' % (t, v, tbinfo))
             self.thunks = []
-        finally:
-            self.lock.release()
 
     def __repr__(self):
         return '<select-trigger (%s) at %x>' % (self.kind, positive_id(self))
 
-if os.name == 'posix':
 
-    class trigger(_triggerbase, asyncore.file_dispatcher):
-        kind = "pipe"
+class pipetrigger(_triggerbase, asyncore.file_dispatcher):
+    kind = "pipe"
 
-        def __init__(self):
-            _triggerbase.__init__(self)
-            r, self.trigger = self._fds = os.pipe()
-            asyncore.file_dispatcher.__init__(self, r)
-
-        def _close(self):
-            for fd in self._fds:
+    def __init__(self):
+        _triggerbase.__init__(self)
+        r, self.trigger = self._fds = os.pipe()
+        asyncore.file_dispatcher.__init__(self, r)
+
+        if self.socket.fd != r:
+            # Starting in Python 2.6, the descriptor passed to
+            # file_dispatcher gets duped and assigned to
+            # self.socket.fd. This breals the instantiation semantics and
+            # is a bug imo.  I dount it will get fixed, but maybe
+            # it will. Who knows. For that reason, we test for the
+            # fd changing rather than just checking the Python version.
```

Comment:
Typos: breals -> breaks; dount -> doubt.

Reply:
Good catch, easily fixed.

This was [copied from ZEO](https://github.com/zopefoundation/ZEO/commit/f4cf6a533f7be5c57d34d868440b18ab37353696), which no longer uses triggers. (But it had accumulated some fixes that had yet to be copied back to here.)